extensions/hawkbit-extension-artifact-repository-s3/src/test/java/org/eclipse/hawkbit/artifact/repository/S3RepositoryTest.java

```diff
@@ -0,0 +1,169 @@
+
+        // test
+        storeRandomBytes(rndBytes, knownContentType);
+
+        // verify
+        Mockito.verify(amazonS3Mock).putObject(eq(s3Properties.getBucketName()), eq(knownSHA1),
+                inputStreamCaptor.capture(), objectMetaDataCaptor.capture());
+
+        final ObjectMetadata recordedObjectMetadata = objectMetaDataCaptor.getValue();
+        assertThat(recordedObjectMetadata.getContentType()).isEqualTo(knownContentType);
+        assertThat(recordedObjectMetadata.getContentMD5()).isNotNull();
+        assertThat(recordedObjectMetadata.getContentLength()).isEqualTo(rndBytes.length);
+    }
+
+    @Test
+    @Description("Verifies that the amazonS3 client is called to retrieve the correct artifact from S3 and the mapping to the DBArtifact is correct")
+    public void getArtifactBySHA1Hash() {
+        final String knownSHA1Hash = "da39a3ee5e6b4b0d3255bfef95601890afd80709";
+        final long knownContentLength = 100;
+        final String knownContentType = "application/octet-stream";
+        final String knownMd5 = "098f6bcd4621d373cade4e832627b4f6";
+        final String knownMdBase16 = BaseEncoding.base16().lowerCase().encode(knownMd5.getBytes());
+        final String knownMd5Base64 = BaseEncoding.base64().encode(knownMd5.getBytes());
+
+        when(amazonS3Mock.getObject(any(GetObjectRequest.class))).thenReturn(s3ObjectMock);
+        when(s3ObjectMock.getObjectMetadata()).thenReturn(s3ObjectMetadataMock);
+        when(s3ObjectMetadataMock.getContentLength()).thenReturn(knownContentLength);
+        when(s3ObjectMetadataMock.getETag()).thenReturn(knownMd5Base64);
+        when(s3ObjectMetadataMock.getContentType()).thenReturn(knownContentType);
+
+        // test
+        final DbArtifact artifactBySha1 = s3RepositoryUnderTest.getArtifactBySha1(knownSHA1Hash);
+
+        // verify
+        assertThat(artifactBySha1.getArtifactId()).isEqualTo(knownSHA1Hash);
+        assertThat(artifactBySha1.getContentType()).isEqualTo(knownContentType);
+        assertThat(artifactBySha1.getSize()).isEqualTo(knownContentLength);
+        assertThat(artifactBySha1.getHashes().getSha1()).isEqualTo(knownSHA1Hash);
+        assertThat(artifactBySha1.getHashes().getMd5()).isEqualTo(knownMdBase16);
+    }
+
+    @Test
+    @Description("Verifies that the amazonS3 client is not called to put the object to S3 due the artifact already exists on S3")
+    public void artifactIsNotUploadedIfAlreadyExists() throws NoSuchAlgorithmException, IOException {
+        final byte[] rndBytes = randomBytes();
+        final String knownSHA1 = getSha1OfRandomBytes(rndBytes);
+        final String knownContentType = "application/octet-stream";
+
+        when(amazonS3Mock.doesObjectExist(s3Properties.getBucketName(), knownSHA1)).thenReturn(true);
+
+        // test
+        storeRandomBytes(rndBytes, knownContentType);
+
+        // verify
+        Mockito.verify(amazonS3Mock, never()).putObject(eq(s3Properties.getBucketName()), eq(knownSHA1),
```

Comment:
Life can be easier: `verifyNoMoreInteractions(ignoreStubs(amazonS3Mock))` :wink: 
But this is also fine